pkg/logs/logs.go

```diff
@@ -29,103 +35,130 @@ func NewLogsClient(kubernetesClient kclient.ClientInterface) *LogsClient {
 
 var _ Client = (*LogsClient)(nil)
 
-func (o *LogsClient) GetLogsForMode(mode string, componentName string, namespace string, follow bool) ([]map[string]io.ReadCloser, error) {
+func (o *LogsClient) GetLogsForMode(
+	mode string,
+	componentName string,
+	namespace string,
+	follow bool,
+	out io.Writer,
+) (Events, error) {
+	events := Events{
+		Logs: make(chan map[string]interface{}),
+		Err:  make(chan error),
+		Done: make(chan struct{}),
+	}
+
+	go o.getLogsForMode(events, mode, componentName, namespace, follow, out)
+	return events, nil
+}
+
+func (o *LogsClient) getLogsForMode(
+	events Events,
+	mode string,
+	componentName string,
+	namespace string,
+	follow bool,
+	out io.Writer,
+) {
 	var selector string
-	logs := []map[string]io.ReadCloser{}
-	unknownMode := true
+	podChan := make(chan corev1.Pod) // grab the logs of the pod put on this channel
+	errChan := make(chan error)
+	doneChan := make(chan struct{}) // because populating doneChan directly would cause odo logs to exit prematurely.
+
+	go func() {
+		// this go routine gets the logs of the pods put on the podChan
+		for {
+			select {
+			case pod := <-podChan:
+				for _, container := range pod.Spec.Containers {
+					containerLogs, err := o.kubernetesClient.GetPodLogs(pod.Name, container.Name, follow)
+					if err != nil {
+						fmt.Fprintf(out, "failed to get logs for container %s; error: %v", container.Name, err)
+					}
+					events.Logs <- map[string]interface{}{"name": container.Name, "logs": containerLogs}
```

Comment:
This will get executed even if `err != nil`, potentially causing `containerLogs` to added as `nil` in the map; and later in the CLI, the value would be type-asserted as an `io.ReadCloser`, which might cause this to panic, I think, no?

Reply:
Ah, nice catch! 👍🏾

Reply:
Addressed.